zetta_utils/mazepa_layer_processing/segmentation/masks/affinities.py

```diff
@@ -0,0 +1,163 @@
+from __future__ import annotations
+
+from typing import Sequence
+
+import attrs
+import torch
+
+from zetta_utils import builder, mazepa, tensor_ops
+from zetta_utils.geometry.vec import Vec3D
+from zetta_utils.layer.volumetric.index import VolumetricIndex
+from zetta_utils.layer.volumetric.layer import VolumetricLayer
+
+
+@builder.register("AdjustAffinitiesOp")
+@mazepa.taskable_operation_cls
+@attrs.frozen
+class AdjustAffinitiesOp:
+    crop_pad: Sequence[int] = (0, 0, 0)
+
+    def get_input_resolution(  # pylint: disable=no-self-use
+        self, dst_resolution: Vec3D[float]
+    ) -> Vec3D[float]:
+        return dst_resolution
+
+    def with_added_crop_pad(self, crop_pad: Vec3D[int]) -> AdjustAffinitiesOp:
+        return attrs.evolve(self, crop_pad=Vec3D[int](*self.crop_pad) + crop_pad)
+
+    def __call__(
+        self,
+        idx: VolumetricIndex,
+        dst: VolumetricLayer,
+        aff_layer: VolumetricLayer,
+        aff_backup_layer: VolumetricLayer,
+        blackout_mask_layer: VolumetricLayer,
+        snap_mask_layer: VolumetricLayer,
+        threshold_mask_layer: VolumetricLayer,
+        threshold_value: float = 0.85,
+        fill_value: float = 0,
+        rework_mode: bool = False,
+        rework_mask_layer: VolumetricLayer | None = None,
+    ) -> None:
+        """
+        Adjust affinities using masks, making sure to backup any adjusted affinities.
+
+        :param idx: VolumetricIndex for the operation
+        :param src: layer with affinities
+        :param dst: layer where adjusted affinities will be written sparsely
+        :param aff_backup_layer: layer where original affinities will be stored
+            if chunk is adjusted
+        :param aff_backup_layer: layer where original affinities will be stored
+            if chunk is adjusted
+        :param blackout_mask_layer: one-hot mask to black out all affinities
+        :param snap_mask_layer: one-hot mask for which segments cannot span its
+            boundary ("snaps" objects at its boundary)
+        :param threshold_mask_layer: one-hot mask which will adjust affinities below
+            `threshold_value`
+        :param threshold_value: value for thresholding
+        :param fill_value: value to set the adjusted affinities
+        :param rework_mode: bool to use rework_mask_layer
+        :param rework_mask_layer: one-hot mask of locations that need to be reworked
+        """
+        idx_padded = idx.padded(self.crop_pad)
+        if rework_mode and (rework_mask_layer is not None):
+            rework_mask = rework_mask_layer[idx_padded]
+            if (rework_mask != 0).sum().item() == 0:
+                return
```

Comment:
Should handle case where ``rework_mode`` is set and ``rework_mask_layer`` is ``None``.

Reply:
If `rework_mode=True` but there's no `rework_mask_layer`, then don't you think it should just proceed as normal? Effectively, that means the `rework_mask_layer` is one-hot everywhere.

Reply:
That makes sense.